crates/ruff/src/directives.rs

```diff
@@ -95,6 +95,31 @@ fn extract_noqa_line_for(lxr: &[LexResult], locator: &Locator, indexer: &Indexer
                 break;
             }
 
+            // F-strings can be multi-line even without triple quotes. For example,
+            //
+            // ```python
+            // f"foo {
+            //    x
+            //        *
+            //             y
+            // }"
+            // ```
+            //
+            // We expect `noqa` directives on the last line of the f-string.
+            Tok::FStringEnd => {
+                // SAFETY: If we're at the end of a f-string, the indexer must have a
+                // corresponding f-string range for it.
+                let fstring_start = indexer
+                    .fstring_ranges()
+                    .outermost(range.start())
+                    .unwrap()
+                    .start();
+                string_mappings.push(TextRange::new(
+                    locator.line_start(fstring_start),
+                    range.end(),
+                ));
+            }
```

Comment:
I wonder if we could build the fstring ranges directly from using the `fstring_ranges` instead by e.g. using `values` and filtering out overlapping ranges (should be adjacent)

Reply:
Good point. This is actually how it should be done as otherwise the nested f-strings weren't handled properly.